### ISSUE TITLE
EcsRunLauncher tasks fail to start with CLI error

### ISSUE THREAD

**Issue by clayton-cc** [dagster-io/dagster#9562]
### Dagster version

dagster, version 1.0.6

### What's the issue?

When attempting to launch a run using the EcsRunLauncher class, ECS tasks are outputting an error from the injected command from Dagster. From the ECS console, Dagster is sending the command:

```
["dagster","api","execute_run","<large JSON string>"]
```

In the logs I see the Dagster CLI complaining about the input command:

```
2022-08-31 08:56:18 Usage: dagster [OPTIONS] COMMAND [ARGS]...
2022-08-31 08:56:18 CLI tools for working with Dagster.
2022-08-31 08:56:18 Options:
2022-08-31 08:56:18 -v, --version Show the version and exit.
2022-08-31 08:56:18 -h, --help Show this message and exit.
2022-08-31 08:56:18 Commands:
2022-08-31 08:56:18 asset Commands for working with Dagster assets.
2022-08-31 08:56:18 debug Commands for debugging Dagster job runs.
2022-08-31 08:56:18 instance Commands for working with the current Dagster instance.
...
```

If I copy and paste the `large JSON string` and run the command via the container locally with `docker run <image> dagster api execute_run <large JSON string>` it can at least start the task.

### What did you expect to happen?

When the ECS task starts, I would expect the dagster run to at least be started and not error on parsing the command from the CLI

### How to reproduce?

As part of my `dagster.yaml` I have the `EcsRunLauncher` defined:

```yaml
run_launcher:
  module: "dagster_aws.ecs"
  class: "EcsRunLauncher"
  config:
    task_definition: <task_definition_arn>
    container_name: <task_container_name>
    include_sidecars: true
```

I have a simple job to keep a task busy for ~60s

```python
import time

from dagster import graph, op, repository


@op
def my_op():
    start = time.time()
    i = 0
    while time.time() - start < 60:
        i = i + 1
        if i % 1000 == 0:
            print(i)
    return True


@graph
def my_graph():
    my_op()


my_job = my_graph.to_job()


@repository
def busy():
    return [my_job]
```

I can start the job from the UI and it can run fine using the default run launcher, but when switching to the `EcsRunLauncher` I am getting errors starting the job.

I have two other containers running `dagit` and the `dagster-daemon` in a ECS separate task

### Deployment type

Other

### Deployment details

This repo https://github.com/datarootsio/terraform-aws-ecs-dagster serves as the basis for our configuration currently which we are setting up to compare dagster against other job tools

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by jamiedemaria** [comment]
@johannkm do you mind taking a look at this?

**Comment by johannkm** [comment]
Very odd. Dagster `1.0.6` is what's installed in the image for the run?

**Comment by clayton-cc** [comment]
Yes, the Dockerfile for the container in the task definition is just:

```Dockerfile
FROM python:3.9-slim

ENV DAGSTER_HOME=/opt/

WORKDIR $DAGSTER_HOME

# Added a copy of the job files, workspace and dagster yaml just in case
COPY ./<dagster_files> ./

RUN python -m pip install --upgrade pip && python -m pip install dagster awscli
```

If I build/run the container locally:

```bash
docker build -t dagster-runner:latest .
docker run dagster-runner:latest dagster --version
> dagster, version 1.0.6
```

For `dagit` and `dagster-daemon` we are using the `dagster/k8s-dagit-example:1.0.6` image. Originally we were using the `latest` tag on that which seemed to be pointing at `0.15.9` but moved to make sure they were all `1.0.6`

**Comment by clayton-cc** [comment]
I did some debugging and I am able to _manually_ start a task by altering the underlying task definition command JSON to be:

```
...
    "command": [
        "/bin/bash",
        "-c",
        "dagster api execute_run '<large JSON string>'"
    ],
...
```

when playing around manually with changing the command override via the UI, it would always break on any commas, even if they were within a string. The first chunk of `<large JSON string>` is:

```
'{\"__class__\": \"ExecuteRunArgs\", \"instance_ref\": {\"__class__\": \"InstanceRef\", ... }'
```

And in the UI for starting a task I try to override the command to be what should be equivalent to the command I edited within the JSON `/bin/bash, -c, dagster api execute_run '<large JSON string'` there is an error parsing the command which points to the information in the `large JSON string` between the first and second comma, as if it is breaking on the comma despite what should be parsed by ECS as a string

```
2022-09-01 13:23:11\"instance_ref\": {\"__class__\": \"InstanceRef\": -c: line 1: unexpected EOF while looking for matching `''
2022-09-01 13:23:11\"instance_ref\": {\"__class__\": \"InstanceRef\": -c: line 2: syntax error: unexpected end of file
```

If I have some more time to debug, I may try to pull down the full dagster code locally, as I'm wondering if changing [this section](https://github.com/dagster-io/dagster/blob/e1adadd33cecf01474ec1c232b5b52b68ab1210a/python_modules/dagster/dagster/_grpc/types.py#L114-L119) to match the JSON edited above would work?

```python
    def get_command_args(self) -> List[str]:
        # If my containers entrypoint is /bin/bash -c, should that be returned here?
        return _get_entry_point(self.pipeline_origin) + [
            f"api execute_run {serialize_dagster_namedtuple(self)}",
        ]
```

**Comment by johannkm** [comment]
Sorry for the delay here- we will take a look this week

**Comment by clayton-cc** [comment]
Any updates? Should I try latest? Ty!

**Comment by jmsanders** [comment]
Do you have a `CMD` or `ENTRYPOINT` defined in your task definition? You might be running into: https://aws.amazon.com/blogs/opensource/demystifying-entrypoint-cmd-docker/

So Dagster is indeed sending the command:

```
["dagster","api","execute_run","<large JSON string>"]
```

but ECS might then be array concatenating it with whatever is in `CMD` or `ENTRYPOINT` in your task definition to instead run:

```
["<another command>", "dagster", "api", "execute_run", <large JSON string>"]
```

At least when we've seen similar issues in the past, that has been the culprit.

I'm a little hesitant to change the default command behavior for the launcher without knowing if this is specific to your custom task definition but I'm curious to know what your task definition looks like so we can either provide better docs/error messages or make the launcher compatible with this kind of customization.

**Comment by clayton-cc** [comment]
Entrypoint in the task definition seems to have been the culprit! I've ran into the issue before as well in a docker-specific context, which is why we tried running with both `sh -c` and `/bin/bash -c`, but I never thought to _remove_ it 🤦 

I think a note [within this section ](https://docs.dagster.io/deployment/guides/aws#launching-runs-in-ecs)about the entrypoint would be beneficial, as our standard terraform setup for ECS uses `sh -c` for all task definitions and a command syntax of `/bin/bash -c \"${var.command}\"`

**Comment by jmsanders** [comment]
🙌 Awesome

I'll add a note to the docs (unless you'd like to open the PR).

Out of curiosity, what was the `ENTRYPOINT`? I couldn't quite reproduce the same set of log messages (other than by running `dagster -h` which I don't think was what's happening here) so I'm curious about what initial command(s) caused the CLI to display the help message.

**Comment by clayton-cc** [comment]
Simply adding `sh, -c` in the task definition as the entrypoint seems to trigger the same error.

![image](https://user-images.githubusercontent.com/97914759/192596722-86d4e141-2713-4fab-b572-b9d7145c09fe.png)
![image](https://user-images.githubusercontent.com/97914759/192596791-30255b84-8279-4486-8a61-eb01948383a5.png)
![image](https://user-images.githubusercontent.com/97914759/192596932-db9b5a87-00a0-47fc-8098-2cddc7e4720e.png)

I _also_ noticed when testing this, that I tried to update my `dagster.yaml` config to point to an older task revision, and it pulled the latest one 👀 

In `dagster.yaml`:

```yaml
run_launcher:
  module: "dagster_aws.ecs"
  class: "EcsRunLauncher"
  config:
    task_definition: "arn:aws:ecs:<region><account_id>:task-definition/dagster-ecs-dagster-test-manual:6"
    container_name: "dagit"
    include_sidecars: true
```

Yet the task that launched was:

![image](https://user-images.githubusercontent.com/97914759/192597307-e4e93624-59e2-4b13-87ad-0f2f9745441d.png)

I thought I couldn't recreate the error, because I tried referencing out `task-definition:6` which was an old version with the `sh -c` as the entrypoint, but it launched with the `task-definition:11` which was the one I had just created to test with no entrypoint or command syntax. I created the 12th version with `sh -c` re-added and recreated the error in the screenshots above

**Comment by cmpadden** [comment]
Closing as the ECS task builder now explicitly strips entryPoint before applying the Dagster command, with an inline note that ECS concatenates entryPoint and container overrides and breaks the command otherwise.
